### PR TITLE
[Feature] Metrics structure

### PR DESCRIPTION
**Pull request by sadroeck**
## Motivation

While micro-benchmarks are relevant in order to optimize a service as a single node, metrics are essential in order to analyze a distributed system. This PR proposes an initial step towards a more structural approach to system metrics.
This allows for a more targeted approach to optimizing network behavior at scale.

Using the same structural definition, the metrics should be queryable using both Prometheus & the `json-RPC` API.

## Approach

- [X] Metrics are namespaced in their relevant domain
  - Note: I feel the current categorization is somewhat strange, so this can definitely be improved
- [X] Use type-safe metric type Wrappers, e.g. `Counter`, `DiscreteGauge`, etc 
- [X] Move metrics definition, background handling & call-site macros to the `snarkos-metrics` crate, to prevent duplication of crate includes
- [X] Deprecate the `snarkos-metrics` binary
- [X] Move `json-RPC` `Stat` type definitions to `snarkos-metrics` to prevent intermediate type conversions between the internal declaration & the expected API type.

### Future/nice-to-haves

- Consolidate metric declaration, categorization & handling using macros.
- Consolidate metrics exporters, allowing a prometheus-compatible export without drain-only behavior, yet allowing a fixed time window for histograms.
- Allow metrics aggregation over the mining network

## Test Plan

Testing is currently limited to running a prometheus instance, manually scraping the prometheus endpoint or querying the `json-RPC` API. Not sure if it's worth the effort to add automated testing here.

## Related PRs

N/A

### CODE REVIEW

**Comment by ljedrz** [comment]
Note: I changed the base branch to `staging`, which is the branch to file PRs against (as your PR is, judging by the commits :+1:).

**Comment by sadroeck** [comment]
> Note: I changed the base branch to `staging`, which is the branch to file PRs against (as your PR is, judging by the commits 👍).
> 
> Edit: actually, it seems that you beat me to it 🙂.

Thank you! Yeah I noticed it a bit too late, force of habit ;-)

File: metrics/prometheus/linux/prometheus.yml [hunk 1]
```diff
@@ -8,6 +8,6 @@ scrape_configs:
     static_configs:
       - targets: ['localhost:9090']
 
-  - job_name: 'service-collector'
+  - job_name: 'mining-node-collector'
```
Comment (ljedrz) [inline metrics/prometheus/linux/prometheus.yml:11]:
```suggestion
  - job_name: 'snarkOS-metrics-collector'
```
metrics can be collected from non-miners as well; this applies to `mac` and `windows` config files

File: metrics/prometheus/windows/prometheus.yml [hunk 2]
```diff
@@ -8,6 +8,6 @@ scrape_configs:
     static_configs:
       - targets: ['host.docker.internal:9090'] # Only works on Docker Desktop for Windows
 
-  - job_name: 'service-collector'
+  - job_name: 'mining-node-collector'
     static_configs:
-      - targets: ['192.168.65.1:8080']
+      - targets: ['192.168.65.1:9000']
```
Comment (ljedrz) [inline metrics/prometheus/windows/prometheus.yml:13]: the port checks out, but what about the IP? shouldn't it be something non-hardcoded?

File: metrics/src/metric_types.rs [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+
+impl Counter {
+    pub(crate) const fn new() -> Self {
+        Self(AtomicU64::new(0))
+    }
+
+    /// Increases the value of the [`Counter`] by a discrete amount
+    #[inline]
+    pub(crate) fn increment(&self, val: u64) {
+        self.0.fetch_add(val, Ordering::SeqCst);
```
Comment (ljedrz) [inline metrics/src/metric_types.rs:14]:
```suggestion
        self.0.fetch_add(val, Ordering::Relaxed);
```
This counter can only ever increase, so `Relaxed` ordering should be sufficient here, no?

**Review by ljedrz — CHANGES_REQUESTED** [review]
There are some very nice improvements to the overall structure in this PR :+1:; there are just a few details that need to be addressed:
- a `cargo fmt` pass using the `nightly` toolchain
- the RPC snapshots currently still need to be available even with `features=prometheus` _and_ some of them need to have "overrides" (as they currently do) that allow us to know if the `prometheus` metrics are valid, as not all of them currently are, especially the gauges; https://github.com/AleoHQ/snarkOS/issues/782 should actually now be fixable due to the new structure, but https://github.com/AleoHQ/snarkOS/issues/869 and https://github.com/AleoHQ/snarkOS/issues/877 are recent regressions; until we are confident that the `metrics`-driven values are perfectly correct, we need some of the currently existing overrides - they can be exposed as another RPC if need be
- the `outbound` metrics seem to have disappeared? I don't see them in `grafana` when running this branch locally, and I don't see them anywhere in the diff's "green"

I'm attaching the JSON model for my `grafana` dashboard, it might be useful:

<details><summary>grafana JSON model</summary>

```
{
  "annotations": {
    "list": [
      {
        "builtIn": 1,
        "datasource": "-- Grafana --",
        "enable": true,
        "hide": true,
        "iconColor": "rgba(0, 211, 255, 1)",
        "name": "Annotations & Alerts",
        "type": "dashboard"
      }
    ]
  },
  "editable": true,
  "gnetId": null,
  "graphTooltip": 0,
  "id": 3,
  "links": [],
  "panels": [
    {
      "datasource": null,
      "description": "",
      "fieldConfig": {
        "defaults": {
          "color": {
            "mode": "palette-classic"
          },
          "custom": {
            "axisLabel": "",
            "axisPlacement": "auto",
            "barAlignment": 0,
            "drawStyle": "line",
            "fillOpacity": 10,
            "gradientMode": "none",
            "hideFrom": {
              "graph": false,
              "legend": false,
              "tooltip": false,
              "viz": false
            },
            "lineInterpolation": "smooth",
            "lineStyle": {
              "fill": "solid"
            },
            "lineWidth": 1,
            "pointSize": 5,
            "scaleDistribution": {
              "type": "linear"
            },
            "showPoints": "never",
            "spanNulls": true,
            "stacking": {
              "group": "A",
              "mode": "none"
            },
            "thresholdsStyle": {
              "mode": "off"
            }
          },
          "mappings": [],
          "min": 0,
          "thresholds": {
            "mode": "absolute",
            "steps": [
              {
                "color": "green",
                "value": null
              },
              {
                "color": "red",
                "value": 80
              }
            ]
          },
          "unit": "short"
        },
        "overrides": []
      },
      "gridPos": {
        "h": 10,
        "w": 11,
        "x": 0,
        "y": 0
      },
      "id": 6,
      "options": {
        "graph": {},
        "legend": {
          "calcs": [
            "mean",
            "max",
            "min",
            "last"
          ],
          "displayMode": "table",
          "placement": "bottom"
        },
        "tooltip": {
          "mode": "multi"
        }
      },
      "pluginVersion": "7.5.6",
      "targets": [
        {
          "exemplar": true,
          "expr": "snarkos_connections_connected_total",
          "hide": false,
          "interval": "",
          "legendFormat": "connected",
          "refId": "D"
        },
        {
          "exemplar": true,
          "expr": "snarkos_connections_disconnected_total",
          "hide": false,
          "interval": "",
          "legendFormat": "disconnected",
          "refId": "E"
        },
        {
          "exemplar": true,
          "expr": "snarkos_connections_connecting_total",
          "hide": false,
          "interval": "",
          "legendFormat": "connecting",
          "refId": "F"
        }
      ],
      "timeFrom": null,
      "timeShift": null,
      "title": "states of current connections",
      "type": "timeseries"
    },
    {
      "datasource": null,
      "description": "",
      "fieldConfig": {
        "defaults": {
          "color": {
            "mode": "palette-classic"
          },
          "custom": {
            "hideFrom": {
              "legend": false,
              "tooltip": false,
              "viz": false
            }
          },
          "mappings": []
        },
        "overrides": []
      },
      "gridPos": {
        "h": 10,
        "w": 7,
        "x": 11,
        "y": 0
      },
      "id": 2,
      "options": {
        "displayLabels": [],
        "legend": {
          "displayMode": "list",
          "placement": "right",
          "values": [
            "percent"
          ]
        },
        "pieType": "pie",
        "reduceOptions": {
          "calcs": [
            "last"
          ],
          "fields": "",
          "values": false
        },
        "text": {},
        "tooltip": {
          "mode": "single"
        }
      },
      "pluginVersion": "7.5.6",
      "targets": [
        {
          "exemplar": true,
          "expr": "snarkos_inbound_blocks_total",
          "hide": false,
          "interval": "",
          "legendFormat": "Block",
          "refId": "blocks"
        },
        {
          "exemplar": true,
          "expr": "snarkos_inbound_getblocks_total",
          "hide": false,
          "interval": "",
          "legendFormat": "GetBlocks",
          "refId": "getblocks"
        },
        {
          "exemplar": true,
          "expr": "snarkos_inbound_getmemorypool_total",
          "hide": false,
          "interval": "",
          "legendFormat": "GetMemoryPool",
          "refId": "getmemorypool"
        },
        {
          "exemplar": true,
          "expr": "snarkos_inbound_getpeers_total",
          "hide": false,
          "interval": "",
          "legendFormat": "GetPeers",
          "refId": "getpeers"
        },
        {
          "exemplar": true,
          "expr": "snarkos_inbound_getsync_total",
          "hide": false,
          "interval": "",
          "legendFormat": "GetSync",
          "refId": "getsync"
        },
        {
          "exemplar": true,
          "expr": "snarkos_inbound_memorypool_total",
          "hide": false,
          "interval": "",
          "legendFormat": "MemoryPool",
          "refId": "memorypool"
        },
        {
          "exemplar": true,
          "expr": "snarkos_inbound_peers_total",
          "hide": false,
          "interval": "",
          "legendFormat": "Peers",
          "refId": "peers"
        },
        {
          "exemplar": true,
          "expr": "snarkos_inbound_pings_total",
          "hide": false,
          "interval": "",
          "legendFormat": "Ping",
          "refId": "pings"
        },
        {
          "exemplar": true,
          "expr": "snarkos_inbound_pongs_total",
          "hide": false,
          "instant": false,
          "interval": "",
          "legendFormat": "Pong",
          "refId": "pongs"
        },
        {
          "exemplar": true,
          "expr": "snarkos_inbound_syncs_total",
          "hide": false,
          "interval": "",
          "legendFormat": "Sync",
          "refId": "sync"
        },
        {
          "exemplar": true,
          "expr": "snarkos_inbound_syncblocks_total",
          "hide": false,
          "interval": "",
          "legendFormat": "SyncBlock",
          "refId": "syncblock"
        },
        {
          "exemplar": true,
          "expr": "snarkos_inbound_transactions_total",
          "hide": false,
          "interval": "",
          "legendFormat": "Transaction",
          "refId": "transaction"
        }
      ],
      "timeFrom": null,
      "timeShift": null,
      "title": "inbound message breakdown",
      "type": "piechart"
    },
    {
      "datasource": null,
      "description": "",
      "fieldConfig": {
        "defaults": {
          "color": {
            "mode": "palette-classic"
          },
          "custom": {
            "hideFrom": {
              "legend": false,
              "tooltip": false,
              "viz": false
            }
          },
          "mappings": []
        },
        "overrides": []
      },
      "gridPos": {
        "h": 10,
        "w": 6,
        "x": 18,
        "y": 0
      },
      "id": 16,
      "options": {
        "displayLabels": [],
        "legend": {
          "displayMode": "list",
          "placement": "right",
          "values": [
            "percent"
          ]
        },
        "pieType": "pie",
        "reduceOptions": {
          "calcs": [
            "lastNotNull"
          ],
          "fields": "",
          "values": false
        },
        "text": {},
        "tooltip": {
          "mode": "single"
        }
      },
      "pluginVersion": "7.5.6",
      "targets": [
        {
          "exemplar": true,
          "expr": "snarkos_inbound_all_successes_total",
          "interval": "",
          "legendFormat": "valid",
          "refId": "A"
        },
        {
          "exemplar": true,
          "expr": "snarkos_inbound_all_failures_total",
          "hide": false,
          "interval": "",
          "legendFormat": "invalid",
          "refId": "B"
        }
      ],
      "title": "all processed inbound messages",
      "type": "piechart"
    },
    {
      "datasource": null,
      "fieldConfig": {
        "defaults": {
          "color": {
            "mode": "palette-classic"
          },
          "custom": {
            "axisLabel": "",
            "axisPlacement": "auto",
            "barAlignment": 0,
            "drawStyle": "line",
            "fillOpacity": 10,
            "gradientMode": "none",
            "hideFrom": {
              "graph": false,
              "legend": false,
              "tooltip": false,
              "viz": false
            },
            "lineInterpolation": "smooth",
            "lineStyle": {
              "fill": "solid"
            },
            "lineWidth": 1,
            "pointSize": 5,
            "scaleDistribution": {
              "type": "linear"
            },
            "showPoints": "never",
            "spanNulls": true,
            "stacking": {
              "group": "A",
              "mode": "none"
            },
            "thresholdsStyle": {
              "mode": "off"
            }
          },
          "mappings": [],
          "thresholds": {
            "mode": "absolute",
            "steps": [
              {
                "color": "green",
                "value": null
              },
              {
                "color": "#EAB839",
                "value": 500
              },
              {
                "color": "red",
                "value": 1000
              }
            ]
          },
          "unit": "short"
        },
        "overrides": []
      },
      "gridPos": {
        "h": 10,
        "w": 5,
        "x": 0,
        "y": 10
      },
      "id": 10,
      "options": {
        "graph": {},
        "legend": {
          "calcs": [
            "min",
            "max",
            "mean"
          ],
          "displayMode": "table",
          "placement": "bottom"
        },
        "tooltip": {
          "mode": "multi"
        }
      },
      "pluginVersion": "7.5.6",
      "targets": [
        {
          "exemplar": true,
          "expr": "snarkos_queues_inbound_total",
          "instant": false,
          "interval": "",
          "legendFormat": "inbound",
          "refId": "A"
        },
        {
          "exemplar": true,
          "expr": "snarkos_queues_outbound_total",
          "hide": false,
          "interval": "",
          "legendFormat": "outbound",
          "refId": "B"
        }
      ],
      "timeFrom": null,
      "timeShift": null,
      "title": "queued messages",
      "type": "timeseries"
    },
    {
      "datasource": null,
      "description": "",
      "fieldConfig": {
        "defaults": {
          "color": {
            "mode": "palette-classic"
          },
          "custom": {
            "hideFrom": {
              "legend": false,
              "tooltip": false,
              "viz": false
            }
          },
          "mappings": []
        },
        "overrides": []
      },
      "gridPos": {
        "h": 10,
        "w": 6,
        "x": 5,
        "y": 10
      },
      "id": 4,
      "options": {
        "displayLabels": [],
        "legend": {
          "displayMode": "list",
          "placement": "right",
          "values": [
            "percent"
          ]
        },
        "pieType": "pie",
        "reduceOptions": {
          "calcs": [
            "lastNotNull"
          ],
          "fields": "",
          "values": false
        },
        "text": {},
        "tooltip": {
          "mode": "single"
        }
      },
      "pluginVersion": "7.5.6",
      "targets": [
        {
          "exemplar": true,
          "expr": "snarkos_outbound_all_successes_total",
          "interval": "",
          "legendFormat": "delivered",
          "refId": "A"
        },
        {
          "exemplar": true,
          "expr": "snarkos_outbound_all_failures_total",
          "hide": false,
          "interval": "",
          "legendFormat": "lost",
          "refId": "B"
        }
      ],
      "timeFrom": null,
      "timeShift": null,
      "title": "outbound messages",
      "type": "piechart"
    },
    {
      "datasource": null,
      "fieldConfig": {
        "defaults": {
          "color": {
            "mode": "palette-classic"
          },
          "custom": {
            "hideFrom": {
              "legend": false,
              "tooltip": false,
              "viz": false
            }
          },
          "mappings": []
        },
        "overrides": []
      },
      "gridPos": {
        "h": 10,
        "w": 7,
        "x": 11,
        "y": 10
      },
      "id": 8,
      "options": {
        "displayLabels": [],
        "legend": {
          "displayMode": "list",
          "placement": "right",
          "values": [
            "percent"
          ]
        },
        "pieType": "pie",
        "reduceOptions": {
          "calcs": [
            "lastNotNull"
          ],
          "fields": "",
          "values": false
        },
        "text": {},
        "tooltip": {
          "mode": "single"
        }
      },
      "pluginVersion": "7.5.6",
      "targets": [
        {
          "exemplar": true,
          "expr": "snarkos_handshakes_successes_init_total",
          "interval": "",
          "legendFormat": "successes as inititiator",
          "refId": "A"
        },
        {
          "exemplar": true,
          "expr": "snarkos_handshakes_failures_init_total",
          "hide": false,
          "interval": "",
          "legendFormat": "failures as initiator",
          "refId": "B"
        },
        {
          "exemplar": true,
          "expr": "snarkos_handshakes_successes_resp_total",
          "hide": false,
          "interval": "",
          "legendFormat": "successes as responer",
          "refId": "C"
        },
        {
          "exemplar": true,
          "expr": "snarkos_handshakes_failures_resp_total",
          "hide": false,
          "interval": "",
          "legendFormat": "failures as responder",
          "refId": "D"
        },
        {
          "exemplar": true,
          "expr": "snarkos_handshakes_timeouts_init_total",
          "hide": false,
          "interval": "",
          "legendFormat": "timeouts as initiator",
          "refId": "E"
        },
        {
          "exemplar": true,
          "expr": "snarkos_handshakes_timeouts_resp_total",
          "hide": false,
          "interval": "",
          "legendFormat": "timeouts as responder",
          "refId": "F"
        }
      ],
      "timeFrom": null,
      "timeShift": null,
      "title": "handshakes",
      "type": "piechart"
    },
    {
      "datasource": null,
      "description": "",
      "fieldConfig": {
        "defaults": {
          "color": {
            "mode": "palette-classic"
          },
          "custom": {
            "hideFrom": {
              "legend": false,
              "tooltip": false,
              "viz": false
            }
          },
          "mappings": []
        },
        "overrides": []
      },
      "gridPos": {
        "h": 10,
        "w": 6,
        "x": 18,
        "y": 10
      },
      "id": 14,
      "options": {
        "displayLabels": [],
        "legend": {
          "displayMode": "list",
          "placement": "right",
          "values": [
            "percent"
          ]
        },
        "pieType": "pie",
        "reduceOptions": {
          "calcs": [
            "lastNotNull"
          ],
          "fields": "",
          "values": false
        },
        "text": {},
        "tooltip": {
          "mode": "single"
        }
      },
      "pluginVersion": "7.5.6",
      "targets": [
        {
          "exemplar": true,
          "expr": "snarkos_connections_all_accepted_total",
          "interval": "",
          "legendFormat": "accepted",
          "refId": "A"
        },
        {
          "exemplar": true,
          "expr": "snarkos_connections_all_rejected_total",
          "hide": false,
          "interval": "",
          "legendFormat": "rejected",
          "refId": "B"
        },
        {
          "exemplar": true,
          "expr": "snarkos_connections_all_initiated_total",
          "hide": false,
          "interval": "",
          "legendFormat": "initiated",
          "refId": "C"
        }
      ],
      "timeFrom": null,
      "timeShift": null,
      "title": "all connections",
      "type": "piechart"
    },
    {
      "datasource": null,
      "fieldConfig": {
        "defaults": {
          "color": {
            "fixedColor": "rgba(128, 128, 128, 0)",
            "mode": "fixed"
          },
          "mappings": [],
          "thresholds": {
            "mode": "absolute",
            "steps": [
              {
                "color": "rgb(140, 140, 140)",
                "value": null
              }
            ]
          }
        },
        "overrides": []
      },
      "gridPos": {
        "h": 4,
        "w": 11,
        "x": 0,
        "y": 20
      },
      "id": 12,
      "options": {
        "colorMode": "background",
        "graphMode": "none",
        "justifyMode": "center",
        "orientation": "auto",
        "reduceOptions": {
          "calcs": [
            "last"
          ],
          "fields": "",
          "values": false
        },
        "text": {},
        "textMode": "auto"
      },
      "pluginVersion": "8.0.2",
      "targets": [
        {
          "exemplar": true,
          "expr": "snarkos_misc_block_height_total",
          "format": "time_series",
          "hide": false,
          "instant": false,
          "interval": "",
          "legendFormat": "block height",
          "refId": "A"
        },
        {
          "exemplar": true,
          "expr": "snarkos_misc_blocks_mined_total",
          "instant": false,
          "interval": "",
          "legendFormat": "mined blocks",
          "refId": "B"
        },
        {
          "exemplar": true,
          "expr": "snarkos_misc_duplicate_blocks_total",
          "hide": false,
          "interval": "",
          "legendFormat": "duplicate Blocks",
          "refId": "C"
        },
        {
          "exemplar": true,
          "expr": "snarkos_misc_duplicate_sync_blocks_total",
          "hide": false,
          "interval": "",
          "legendFormat": "duplicate SyncBlocks",
          "refId": "D"
        },
        {
          "exemplar": true,
          "expr": "snarkos_misc_rpc_requests_total",
          "hide": false,
          "interval": "",
          "legendFormat": "RPC requests",
          "refId": "E"
        }
      ],
      "timeFrom": null,
      "timeShift": null,
      "title": "miscellaneous",
      "type": "stat"
    }
  ],
  "refresh": "5s",
  "schemaVersion": 30,
  "style": "dark",
  "tags": [
    "prometheus"
  ],
  "templating": {
    "list": []
  },
  "time": {
    "from": "now-5m",
    "to": "now"
  },
  "timepicker": {},
  "timezone": "",
  "title": "snarkOS node",
  "uid": "PAzNcaCGz",
  "version": 47
}
```

</details>

File: metrics/README.md [hunk 4]
```diff
@@ -6,22 +6,20 @@
 
 ## Development
 
-To start a local instance of Prometheus for development purposes, run:
+To start a local instance of Prometheus for development purposes, run the following command from the current directory:
```
[the remark this answers is not in the document]
Comment (ljedrz) [reply metrics/README.md:9]: Confirmed, I use `systemd` to run both `prometheus` and `grafana`.

File: metrics/prometheus/windows/prometheus.yml (hunk 2 above)
Comment (sadroeck) [reply metrics/prometheus/windows/prometheus.yml:13]: Iirc, `192.168.65.1` is the old "magical" docker IP address that resolves to the host IP (now `host.docker.internal`). However as I didn't have any windows machine/VM around, I just retained this IP.

File: metrics/src/metric_types.rs (hunk 3 above)
Comment (sadroeck) [reply metrics/src/metric_types.rs:14]: You can still get read-after-write data races using `Relaxed` ordering. However I do think we can reduce the guarantees down to `Release`, as snapshot reads are always relaxed. The same for the Gauges.

Comment (ljedrz) [reply metrics/src/metric_types.rs:14]: It's just a nit though; we can be conservative for now and adjust later if need be 👍.

**Comment by sadroeck** [comment]
>     * a `cargo fmt` pass using the `nightly` toolchain
Done! I've coaxed my IDE into using the nightly formatter.

>     * the `outbound` metrics seem to have disappeared? I don't see them in `grafana` when running this branch locally, and I don't see them anywhere in the diff's "green"

I've removed the `register_metrics`, so they will only show up in the prometheus stats once first incremented. The only usage I can see is in `network/src/outbound/outbound.rs`, which is not compiled it seems. Or am I looking at the wrong `outbound` stats here?

> I'm attaching the JSON model for my `grafana` dashboard, it might be useful:
> grafana JSON model

Would you mind if I add this to the repo? Could be useful for others :-)?

>  the RPC snapshots currently still need to be available even with `features=prometheus` _and_ some of them need to
> have "overrides" (as they currently do) that allow us to know if the `prometheus` metrics are valid, as not all of them
> currently are, especially the gauges; #782 should actually now be fixable due to the new structure, and #869 and #877 are 
> recent regressions; until we are confident that the `metrics`-driven values are perfectly correct, we need some of the 
> currently existing overrides - they can be exposed as another RPC if need be

So my `future` proposal here is to consolidate both metric types, removing the `metrics-exporter-prometheus` & just allow exporting the `Stats` in the prometheus format. This would make the `Stats` the single point of truth. As this is also fully statically initialized, there no lazy or conditional initialization anymore, preventing any metrics from being missed at start-up. 
From a mapping perspective (key -> metric), I was planning on doing via registration macros to prevent client-side errors.
That should be sufficient in order to consider the metrics "correct".

File: metrics/prometheus/linux/prometheus.yml (hunk 1 above)
Comment (sadroeck) [reply metrics/prometheus/linux/prometheus.yml]: updated to `snarkos-metrics-collector` for all configs.

File: metrics/src/metric_types.rs (hunk 3 above)
Comment (sadroeck) [reply metrics/src/metric_types.rs]: That's fair. Given the uncertainty towards the correctness of the metrics, being conservative is the better choice. I had already relaxed them down to `Release`, so I'm happy to leave them as-is :-)

**Comment by ljedrz** [comment]
>The only usage I can see is in network/src/outbound/outbound.rs, which is not compiled it seems.

@Protryon any reason why that module wasn't just removed as part of `lockless`? It is, in fact, not compiled, and the outbound metrics are thus not incremented ever; we do want to measure the number of successfully sent messages, however, so we should bring them back, but it is indeed outside of this PR's scope.

>Would you mind if I add this to the repo? Could be useful for others :-)?

I'm ok with this :ok_hand:.

> future proposal (...)

As for the future plans, I agree that the end goal would be to only use "proper" metrics, but right now we are not incrementing/decrementing everything in the right place, so having an alternative way to view the number of connections or block height, even in a "clunky" way (e.g. using `Ledger::get_current_block_height` as opposed to the counterpart metric) is preferable to having unreliable values as the only reference, which can make debugging efforts more difficult. As far as I'm aware, it's currently the aforementioned `outbound` metrics (plus the `OUTBOUND` queue), the block height and the `connections` metrics.

**Comment by sadroeck** [comment]
> > future proposal (...)
> 
> As for the future plans, I agree that the end goal would be to only use "proper" metrics, but right now we are not incrementing/decrementing everything in the right place, so having an alternative way to view the number of connections or block height, even in a "clunky" way (e.g. using `Ledger::get_current_block_height` as opposed to the counterpart metric) is preferable to having unreliable values as the only reference, which can make debugging efforts more difficult. As far as I'm aware, it's currently the aforementioned `outbound` metrics (plus the `OUTBOUND` queue), the block height and the `connections` metrics.

Added the overrides back into 751eba37

**Review by ljedrz — APPROVED** [review]
LGTM :+1:

**Review by niklaslong — APPROVED** [review]

**Pull request merged**